README
This is the README file of python-mcrypt.

Description
-----------

The python mcrypt module provides an interface for the mcrypt library.

Mcrypt is a library which provides a uniform interface to several symmetric
encryption algorithms. It is intended to have a simple interface to access
encryption algorithms in ofb, cbc, cfb, ecb and stream modes. The algorithms it
supports are DES, 3DES, RIJNDAEL, Twofish, IDEA, GOST, CAST-256, ARCFOUR,
SERPENT, SAFER+, and more.

Features
--------

The python-mcrypt module provides the following features:

o  Free software (GNU Lesser General Public License).
o  Support for Python newstyle classes, allowing subclassing.
o  Comprehensive unitttests.
o  Support for many well known and less known algorithms and modes. (*)
o  Support for fixlength algorithm, adjusting encrypted data length
   even for block algorithms and modes.
o  Support for direct file encryption.
o  Complete inline documentation.
o  Check parameters before passing to mcrypt library, protecting
   agains segmentation faults.

(*) mcrypt library features

Disclaimer
----------

Please read the file LICENSE before using python-mcrypt.

New Versions
------------

Information on how to retrieve new versions of python-mcrypt can be
found on http://sourceforge.net/projects/python-mcrypt.

Credits
-------

Please read the file AUTHORS.



NEWS
This is the NEWS file of python-mcrypt.

Version 1.1
	* Fixed endline inside doc strings. This was breaking
	  compilation on newer gcc.

Version 1.0
	* First public release


libmcrypt-2.5.8/src/idea_test.c
/*
 *    Copyright (C) 1998,1999,2000,2002 Nikos Mavroyanopoulos
 *
 *    This program is free software; you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as published by
 *    the Free Software Foundation; either version 2 of the License, or
 *    (at your option) any later version.
 *
 *    This program is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with this program; if not, write to the Free Software
 *    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include "../include/mutils/mcrypt.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* Prints plaintext and ciphertext in hex for all the algorithms */

#define ALGORITHMS_DIR "../modules/algorithms/.libs"
#define MODES_DIR "../modules/modes/.libs"
/* #define ALGORITHMS_DIR NULL
 * #define MODES_DIR NULL
 */
#define TEXT "a small text, just to test the implementation"

int main()
{
	MCRYPT td, td2;
	int i, t, imax;
	int j, jmax, ivsize;
	int x = 0, siz;
	char *text;
	unsigned char *IV;
	unsigned char *key;
	int keysize;
	
    td = mcrypt_module_open("idea", ALGORITHMS_DIR, "cbc", MODES_DIR);
    td2 = mcrypt_module_open("idea", ALGORITHMS_DIR, "cbc", MODES_DIR);

    if (td != MCRYPT_FAILED && td2 != MCRYPT_FAILED) {
        fprintf(stderr, "Created IDEA cipher.\n");

        keysize = mcrypt_enc_get_key_size(td);
        fprintf(stderr, "Cipher key size %d.\n", keysize);

        key = calloc(1, keysize);
        if (key==NULL) exit(1);
					
        for (t=0;t<keysize;t++)
            key[t] = (t % 255) + 13;
					
        ivsize = mcrypt_enc_get_iv_size(td);
        fprintf(stderr, "IV size %d.\n", ivsize);
        if (ivsize>0) {
            IV = calloc( 1, ivsize);
            if (IV==NULL) exit(1);
            for (t=0;t<ivsize;t++)
                IV[t] = (t*2 % 255) + 15;
        }

        if (mcrypt_generic_init( td, key, keysize, IV) < 0) {
            fprintf(stderr, "Failed to Initialize algorithm!\n");
            return -1;
        }

        if (mcrypt_enc_is_block_mode(td)!=0)
            siz = (strlen(TEXT) / mcrypt_enc_get_block_size(td))*mcrypt_enc_get_block_size(td);
        else siz = strlen(TEXT);

        text = calloc( 1, siz);
        if (text==NULL) exit(1);
					
        memmove( text, TEXT, siz);

        mcrypt_generic( td, text, siz);

        if (mcrypt_generic_init( td2, key, keysize, IV) < 0) {
            fprintf(stderr, "Failed to Initialize algorithm!\n");
            return -1;
        }

        mdecrypt_generic( td2, text, siz);
        if ( memcmp( text, TEXT, siz) == 0) {
            printf( "   %s: ok\n", "cbc");
        } else {
            printf( "   %s: failed\n", "cbc");
            x=1;
        }

        mcrypt_generic_deinit(td);
        mcrypt_generic_deinit(td2);
        mcrypt_module_close(td);
        mcrypt_module_close(td2);
        free(text);
        free(key);
        if (ivsize>0) free(IV);
    }

    return 0;
}


libmcrypt-2.5.8/config.h
/* config.h.  Generated from config.h.in by configure.  */
/* config.h.in.  Generated from configure.in by autoheader.  */

/* Define if building universal (internal helper macro) */
/* #undef AC_APPLE_UNIVERSAL_BUILD */

/* Define to 1 if the `closedir' function returns void instead of `int'. */
/* #undef CLOSEDIR_VOID */

/* dlmutex */
/* #undef DLMUTEX */

/* Define to 1 if you have the <argz.h> header file. */
#define HAVE_ARGZ_H 1

/* Define to 1 if you have the `bcopy,,' function. */
/* #undef HAVE_BCOPY__ */

/* Define to 1 if you have the <byteswap.h> header file. */
#define HAVE_BYTESWAP_H 1

/* Define to 1 if you have the `bzero' function. */
#define HAVE_BZERO 1

/* Define to 1 if you have the <dirent.h> header file. */
#define HAVE_DIRENT_H 1

/* Define to 1 if you have the <dlfcn.h> header file. */
#define HAVE_DLFCN_H 1

/* Define to 1 if you have the <endian.h> header file. */
#define HAVE_ENDIAN_H 1

/* Define to 1 if you have the <inttypes.h> header file. */
#define HAVE_INTTYPES_H 1

/* Define to 1 if you have the `dmalloc' library (-ldmalloc). */
/* #undef HAVE_LIBDMALLOC */

/* Define to 1 if you have the <limits.h> header file. */
#define HAVE_LIMITS_H 1

/* Define to 1 if your system has a GNU libc compatible `malloc' function, and
   to 0 otherwise. */
#define HAVE_MALLOC 1

/* Define to 1 if you have the <malloc.h> header file. */
#define HAVE_MALLOC_H 1

/* Define to 1 if you have the `memmove' function. */
#define HAVE_MEMMOVE 1

/* Define to 1 if you have the <memory.h> header file. */
#define HAVE_MEMORY_H 1

/* Define to 1 if you have the `memset' function. */
#define HAVE_MEMSET 1

/* Define to 1 if you have the `mlock' function. */
#define HAVE_MLOCK 1

/* Define to 1 if you have the <ndir.h> header file, and it defines `DIR'. */
/* #undef HAVE_NDIR_H */

/* Define to 1 if you have the `readdir_r' function. */
#define HAVE_READDIR_R 1

/* Define to 1 if you have the `readdir,,' function. */
/* #undef HAVE_READDIR__ */

/* Define to 1 if your system has a GNU libc compatible `realloc' function,
   and to 0 otherwise. */
#define HAVE_REALLOC 1

/* Define to 1 if you have the <stdint.h> header file. */
#define HAVE_STDINT_H 1

/* Define to 1 if you have the <stdlib.h> header file. */
#define HAVE_STDLIB_H 1

/* Define to 1 if you have the `strchr' function. */
#define HAVE_STRCHR 1

/* Define to 1 if you have the `strdup' function. */
#define HAVE_STRDUP 1

/* Define to 1 if you have the <strings.h> header file. */
#define HAVE_STRINGS_H 1

/* Define to 1 if you have the <string.h> header file. */
#define HAVE_STRING_H 1

/* Define to 1 if you have the `strrchr' function. */
#define HAVE_STRRCHR 1

/* Define to 1 if you have the <sys/dir.h> header file, and it defines `DIR'.
   */
/* #undef HAVE_SYS_DIR_H */

/* Define to 1 if you have the <sys/endian.h> header file. */
/* #undef HAVE_SYS_ENDIAN_H */

/* Define to 1 if you have the <sys/mman.h> header file. */
#define HAVE_SYS_MMAN_H 1

/* Define to 1 if you have the <sys/ndir.h> header file, and it defines `DIR'.
   */
/* #undef HAVE_SYS_NDIR_H */

/* Define to 1 if you have the <sys/stat.h> header file. */
#define HAVE_SYS_STAT_H 1

/* Define to 1 if you have the <sys/types.h> header file. */
#define HAVE_SYS_TYPES_H 1

/* Define to 1 if you have the <unistd.h> header file. */
#define HAVE_UNISTD_H 1

/* Define to the sub-directory in which libtool stores uninstalled libraries.
   */
#define LT_OBJDIR ".libs/"

/* Name of package */
#define PACKAGE "libmcrypt"

/* Define to the address where bug reports for this package should be sent. */
#define PACKAGE_BUGREPORT ""

/* Define to the full name of this package. */
#define PACKAGE_NAME ""

/* Define to the full name and version of this package. */
#define PACKAGE_STRING ""

/* Define to the one symbol short name of this package. */
#define PACKAGE_TARNAME ""

/* Define to the home page for this package. */
#define PACKAGE_URL ""

/* Define to the version of this package. */
#define PACKAGE_VERSION ""

/* The size of `unsigned char', as computed by sizeof. */
#define SIZEOF_UNSIGNED_CHAR 1

/* The size of `unsigned int', as computed by sizeof. */
#define SIZEOF_UNSIGNED_INT 4

/* The size of `unsigned long int', as computed by sizeof. */
#define SIZEOF_UNSIGNED_LONG_INT 4

/* The size of `unsigned short int', as computed by sizeof. */
#define SIZEOF_UNSIGNED_SHORT_INT 2

/* Define to 1 if you have the ANSI C header files. */
#define STDC_HEADERS 1

/* dmalloc */
/* #undef USE_DMALLOC */

/* use libltdl */
/* #undef USE_LTDL */

/* Version number of package */
#define VERSION "2.5.8"

/* whether win32 environment */
/* #undef WIN32 */

/* Define WORDS_BIGENDIAN to 1 if your processor stores words with the most
   significant byte first (like Motorola and SPARC, unlike Intel). */
#if defined AC_APPLE_UNIVERSAL_BUILD
# if defined __BIG_ENDIAN__
#  define WORDS_BIGENDIAN 1
# endif
#else
# ifndef WORDS_BIGENDIAN
/* #  undef WORDS_BIGENDIAN */
# endif
#endif

/* Define to empty if `const' does not conform to ANSI C. */
/* #undef const */

/* Define to `__inline__' or `__inline' if that's what the C compiler
   calls it, or to nothing if 'inline' is not supported under any name.  */
#ifndef __cplusplus
/* #undef inline */
#endif

/* Define to rpl_malloc if the replacement function should be used. */
/* #undef malloc */

/* Define to rpl_realloc if the replacement function should be used. */
/* #undef realloc */


libmcrypt-2.5.8/include/mutils/mcrypt.h
/*
 * Copyright (C) 1998,1999,2002 Nikos Mavroyanopoulos
 *
 * Encryption/decryption library. This library is free software;
 * you can redistribute it and/or modify it under the terms of the
 * GNU Library General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any
 * later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */


#define MCRYPT_API_VERSION 20021217

#define LIBMCRYPT_VERSION "2.5.8"

#ifdef __cplusplus
extern "C" {
#endif
/* Definitions
 */
#define MCRYPT_FAILED 0x0

struct CRYPT_STREAM;
typedef struct CRYPT_STREAM *MCRYPT;

/* generic - high level functions. 
 */
	MCRYPT mcrypt_module_open(char *algorithm,
				  char *a_directory, char *mode,
				  char *m_directory);
	int mcrypt_module_close(MCRYPT td);

	/* returns 0 if the library has not been compiled with
	 * dynamic module support.
 	 */
	int mcrypt_module_support_dynamic(void);

/* returns thread descriptor */

	int mcrypt_generic_init(const MCRYPT td, void *key, int lenofkey,
				void *IV);
	int mcrypt_generic_deinit(const MCRYPT td);
	int mcrypt_generic_end(const MCRYPT td);
	int mdecrypt_generic(MCRYPT td, void *plaintext, int len);
	int mcrypt_generic(MCRYPT td, void *plaintext, int len);

/* extra functions */

	int mcrypt_enc_set_state(MCRYPT td, void *st, int size);
	int mcrypt_enc_get_state(MCRYPT td, void *st, int *size); /* only 
	* for block algorithms and certain modes like cbc
	* ncfb etc.
	*/
	int (mcrypt_enc_self_test) (MCRYPT);
	int (mcrypt_enc_get_block_size) (MCRYPT);
	int (mcrypt_enc_get_iv_size) (MCRYPT);
	int (mcrypt_enc_get_key_size) (MCRYPT);

/* If this is a block algorithm returns 1 
 */
	int (mcrypt_enc_is_block_algorithm) (MCRYPT);

/* If the mode operates in blocks returns 1 
 */
	int (mcrypt_enc_is_block_mode) (MCRYPT);

/* If the mode is for block algorithms it returns 1 
 */
	int (mcrypt_enc_is_block_algorithm_mode) (MCRYPT td);
	int mcrypt_enc_mode_has_iv(MCRYPT td);

/* Return a const string containing the name of the algorithm/mode
 */
	char *(mcrypt_enc_get_algorithms_name) (MCRYPT td);
	char *(mcrypt_enc_get_modes_name) (MCRYPT td);

	int *mcrypt_enc_get_supported_key_sizes(MCRYPT td, int *len);


	char **mcrypt_list_algorithms(char *libdir, int *size);
	char **mcrypt_list_modes(char *libdir, int *size);

	/* Frees the memory allocated by the mcrypt_list_xxx() functions. 
	 */
	void mcrypt_free_p(char **p, int size);
	void mcrypt_free(void *ptr);

	/* If mcrypt_xxx functions return an error code, and you supply this
	 * to this function, it will print a human readable message
	 */
	void mcrypt_perror(int err);
	const char* mcrypt_strerror(int err);

	/* Self test for the specified algorithm 
	 */
	int mcrypt_module_self_test(char *algorithm, char *a_directory);

	int mcrypt_module_is_block_algorithm(char *algorithm,
					     char *a_directory);
	int mcrypt_module_is_block_algorithm_mode(char *mode,
						  char *m_directory);
	int mcrypt_module_is_block_mode(char *mode, char *m_directory);

	int mcrypt_module_get_algo_key_size(char *algorithm,
					    char *a_directory);
	int mcrypt_module_get_algo_block_size(char *algorithm,
					      char *a_directory);

	int *mcrypt_module_get_algo_supported_key_sizes(char *algorithm,
							char *a_directory,
							int *len);

	/* Checks the version of the specified module 
	 */
	int mcrypt_module_algorithm_version(char *algorithm,
					    char *a_directory);
	int mcrypt_module_mode_version(char *mode, char *a_directory);


	/* for multithreaded applications: 
	 */
	int mcrypt_mutex_register ( void (*mutex_lock)(void) , 
			void (*mutex_unlock)(void), 
			void (*set_error)(const char*), 
			const char* (*get_error)(void));

	const char *
		mcrypt_check_version( const char *);

	/* These definitions exist in order to ease the access to 
	 * mcrypt_module_init().
	 */

	/* Algorithms */
#define MCRYPT_BLOWFISH		"blowfish"
#define MCRYPT_DES 		"des"
#define MCRYPT_3DES 		"tripledes"
#define MCRYPT_3WAY 		"threeway"
#define MCRYPT_GOST 		"gost"
#define MCRYPT_SAFER_SK64 	"safer-sk64"
#define MCRYPT_SAFER_SK128 	"safer-sk128"
#define MCRYPT_CAST_128 	"cast-128"
#define MCRYPT_XTEA 		"xtea"
#define MCRYPT_RC2	 	"rc2"
#define MCRYPT_TWOFISH 		"twofish"
#define MCRYPT_CAST_256 	"cast-256"
#define MCRYPT_SAFERPLUS 	"saferplus"
#define MCRYPT_LOKI97 		"loki97"
#define MCRYPT_SERPENT 		"serpent"
#define MCRYPT_RIJNDAEL_128 	"rijndael-128"
#define MCRYPT_RIJNDAEL_192 	"rijndael-192"
#define MCRYPT_RIJNDAEL_256 	"rijndael-256"
#define MCRYPT_ENIGMA 		"enigma"
#define MCRYPT_ARCFOUR		"arcfour"
#define MCRYPT_WAKE		"wake"
#define MCRYPT_IDEA		"idea"

	/* Modes */
#define MCRYPT_CBC		"cbc"
#define MCRYPT_ECB		"ecb"
#define MCRYPT_CFB		"cfb"
#define MCRYPT_OFB		"ofb"
#define MCRYPT_nOFB		"nofb"
#define MCRYPT_STREAM		"stream"

#ifdef __cplusplus
}
#endif


libmcrypt-2.5.8/modules/algorithms/idea.c

/*               Minor Changes for endiannes (for mcrypt)                     */
/*               --Nikos                                                      */
/******************************************************************************/
/*                                                                            */
/* I N T E R N A T I O N A L  D A T A  E N C R Y P T I O N  A L G O R I T H M */
/*                                                                            */
/******************************************************************************/
/* Author:       Richard De Moliner (demoliner@isi.ee.ethz.ch)                */
/*               Signal and Information Processing Laboratory                 */
/*               Swiss Federal Institute of Technology                        */
/*               CH-8092 Zuerich, Switzerland                                 */

/* Created:      April 23, 1992                                               */
/* Changes:      November 16, 1993 (support of ANSI-C and C++)                */
/* System:       SUN SPARCstation, SUN acc ANSI-C-Compiler, SUN-OS 4.1.3      */
/******************************************************************************/

/* $Id: idea.c,v 1.3 2000/10/07 18:27:25 nmav Exp $ */

#include "../../lib/libdefs.h"
#include "../../lib/mcrypt_modules.h"
#include "idea.h"

#define mulMod        0x10001	/* 2**16 + 1                                    */
#define ones           0xFFFF	/* 2**16 - 1                                    */

#define _mcrypt_set_key idea_LTX__mcrypt_set_key
#define _mcrypt_encrypt idea_LTX__mcrypt_encrypt
#define _mcrypt_decrypt idea_LTX__mcrypt_decrypt
#define _mcrypt_get_size idea_LTX__mcrypt_get_size
#define _mcrypt_get_block_size idea_LTX__mcrypt_get_block_size
#define _is_block_algorithm idea_LTX__is_block_algorithm
#define _mcrypt_get_key_size idea_LTX__mcrypt_get_key_size
#define _mcrypt_get_supported_key_sizes idea_LTX__mcrypt_get_supported_key_sizes
#define _mcrypt_get_algorithms_name idea_LTX__mcrypt_get_algorithms_name
#define _mcrypt_self_test idea_LTX__mcrypt_self_test
#define _mcrypt_algorithm_version idea_LTX__mcrypt_algorithm_version

/******************************************************************************/
/* Multiplication in the multiplicative group, a = a * b                      */
/* pre:  0 <= a <= 0xFFFF.                                                    */
/*       0 <= b <= 0xFFFF.                                                    */
/* post: 'a' and 'b' have been modified.                                      */
/*       a = a * b; where '*' is multiplication in the multiplicative group.  */
/* note: This implementation of '*' is not complete. To bee complete the      */
/*       result has to bee masked (MUL(a, b); a &= ones;).                    */

#define Mul(a, b)                                                              \
  if (a == 0) a = mulMod - b;                                                  \
  else if (b == 0) a = mulMod - a;                                             \
  else {                                                                       \
    a *= b;                                                                    \
    if ((a & ones) >= (b = a >> 16)) a -= b;                                   \
    else a += mulMod - b;                                                      \
  }				/* Mul */

/******************************************************************************/
/* Encryption and decryption algorithm IDEA. Depending on the value of 'key'  */
/* 'Idea_Crypt' either encrypts or decrypts 'dataIn'. The result is stored    */
/* in 'dataOut'.                                                              */
/* pre:  'dataIn'  contains the plain/cipher-text block.                      */
/*       'key'     contains the encryption/decryption key.                    */
/* post: 'dataOut' contains the cipher/plain-text block.                      */

void _mcrypt_encrypt(IDEA_KEY * key, Idea_Data dataIn)
{
	register word32 x0, x1, x2, x3, t0, t1, t2;
	int round, i = 0;

#ifdef WORDS_BIGENDIAN
	x0 = (word32) (byteswap16(dataIn[0]));
	x1 = (word32) (byteswap16(dataIn[1]));
	x2 = (word32) (byteswap16(dataIn[2]));
	x3 = (word32) (byteswap16(dataIn[3]));
#else
	x0 = (word32) ((dataIn[0]));
	x1 = (word32) ((dataIn[1]));
	x2 = (word32) ((dataIn[2]));
	x3 = (word32) ((dataIn[3]));
#endif

	for (round = Idea_nofRound; round > 0; round--) {
		t1 = (word32) key->Idea_Key[i++];
		x1 += (word32) key->Idea_Key[i++];
		x2 += (word32) key->Idea_Key[i++];
		x2 &= ones;
		t2 = (word32) key->Idea_Key[i++];
		Mul(x0, t1);
		x0 &= ones;
		Mul(x3, t2);
		t0 = (word32) key->Idea_Key[i++];
		t1 = x0 ^ x2;
		Mul(t0, t1);
		t0 &= ones;
		t1 = (word32) key->Idea_Key[i++];
		t2 = ((x1 ^ x3) + t0) & ones;
		Mul(t1, t2);
		t1 &= ones;
		t0 += t1;
		x0 ^= t1;
		x3 ^= t0;
		x3 &= ones;
		t0 ^= x1;
		x1 = x2 ^ t1;
		x2 = t0;
	}
	t0 = (word32) key->Idea_Key[i++];
	Mul(x0, t0);
#ifdef WORDS_BIGENDIAN
	dataIn[0] = byteswap16((word16) (x0 & ones));
	dataIn[1] =
	    byteswap16((word16)
			(((word32) key->Idea_Key[i++] + x2) & ones));
	dataIn[2] =
	    byteswap16((word16)
			(((word32) key->Idea_Key[i++] + x1) & ones));
	t0 = (word32) key->Idea_Key[i];
	Mul(x3, t0);
	dataIn[3] = byteswap16((word16) (x3 & ones));
#else
	dataIn[0] = ((word16) (x0 & ones));
	dataIn[1] = ((word16) (((word32) key->Idea_Key[i++] + x2) & ones));
	dataIn[2] = ((word16) (((word32) key->Idea_Key[i++] + x1) & ones));
	t0 = (word32) key->Idea_Key[i];
	Mul(x3, t0);
	dataIn[3] = ((word16) (x3 & ones));
#endif
}				/* Idea_Crypt */


void _mcrypt_decrypt(IDEA_KEY * key, Idea_Data dataIn)
{
	register word32 x0, x1, x2, x3, t0, t1, t2;
	int round, i = 0;

#ifdef WORDS_BIGENDIAN
	x0 = (word32) (byteswap16(dataIn[0]));
	x1 = (word32) (byteswap16(dataIn[1]));
	x2 = (word32) (byteswap16(dataIn[2]));
	x3 = (word32) (byteswap16(dataIn[3]));
#else
	x0 = (word32) ((dataIn[0]));
	x1 = (word32) ((dataIn[1]));
	x2 = (word32) ((dataIn[2]));
	x3 = (word32) ((dataIn[3]));
#endif

	for (round = Idea_nofRound; round > 0; round--) {
		t1 = (word32) key->Idea_inverted_Key[i++];
		x1 += (word32) key->Idea_inverted_Key[i++];
		x2 += (word32) key->Idea_inverted_Key[i++];
		x2 &= ones;
		t2 = (word32) key->Idea_inverted_Key[i++];
		Mul(x0, t1);
		x0 &= ones;
		Mul(x3, t2);
		t0 = (word32) key->Idea_inverted_Key[i++];
		t1 = x0 ^ x2;
		Mul(t0, t1);
		t0 &= ones;
		t1 = (word32) key->Idea_inverted_Key[i++];
		t2 = ((x1 ^ x3) + t0) & ones;
		Mul(t1, t2);
		t1 &= ones;
		t0 += t1;
		x0 ^= t1;
		x3 ^= t0;
		x3 &= ones;
		t0 ^= x1;
		x1 = x2 ^ t1;
		x2 = t0;
	}
	t0 = (word32) key->Idea_inverted_Key[i++];
	Mul(x0, t0);
#ifdef WORDS_BIGENDIAN
	dataIn[0] = byteswap16((word16) (x0 & ones));
	dataIn[1] =
	    byteswap16((word16)
			(((word32) key->Idea_inverted_Key[i++] + x2) &
			 ones));
	dataIn[2] =
	    byteswap16((word16)
			(((word32) key->Idea_inverted_Key[i++] + x1) &
			 ones));
	t0 = (word32) key->Idea_inverted_Key[i];
	Mul(x3, t0);
	dataIn[3] = byteswap16((word16) (x3 & ones));
#else
	dataIn[0] = ((word16) (x0 & ones));
	dataIn[1] =
	    ((word16)
	     (((word32) key->Idea_inverted_Key[i++] + x2) & ones));
	dataIn[2] =
	    ((word16)
	     (((word32) key->Idea_inverted_Key[i++] + x1) & ones));
	t0 = (word32) key->Idea_inverted_Key[i];
	Mul(x3, t0);
	dataIn[3] = ((word16) (x3 & ones));
#endif
}				/* Idea_Crypt */



/******************************************************************************/
/* Multiplicative Inverse by Extended Stein Greatest Common Divisor Algorithm.*/
/* pre:  0 <= x <= 0xFFFF.                                                    */
/* post: x * MulInv(x) == 1, where '*' is multiplication in the               */
/*                           multiplicative group.                            */
/* static was removed */

word16 MulInv(word16 x)
{

	register sword32 n1, n2, N, a1, a2, b1, b2;

	if (x <= 1)
		return x;
	n1 = N = (sword32) x;
	n2 = mulMod;
	a1 = b2 = 1;
	a2 = b1 = 0;
	do {
		while ((n1 & 1) == 0) {
			if (a1 & 1) {
				if (a1 < 0) {
					a1 += mulMod;
					b1 -= N;
				} else {
					a1 -= mulMod;
					b1 += N;
				}
			}
			n1 >>= 1;
			a1 >>= 1;
			b1 >>= 1;
		}
		if (n1 < n2)
			do {
				n2 -= n1;
				a2 -= a1;
				b2 -= b1;
				if (n2 == 0)
					return (word16) (a1 <
							 0 ? a1 +
							 mulMod : a1);
				while ((n2 & 1) == 0) {
					if (a2 & 1) {
						if (a2 < 0) {
							a2 += mulMod;
							b2 -= N;
						} else {
							a2 -= mulMod;
							b2 += N;
						}
					}
					n2 >>= 1;
					a2 >>= 1;
					b2 >>= 1;
				}
			} while (n1 <= n2);
		n1 -= n2;
		a1 -= a2;
		b1 -= b2;
	} while (n1);
	return (word16) (a2 < 0 ? a2 + mulMod : a2);
}				/* MulInv */

/******************************************************************************/
/* Additive Inverse.                                                          */
/* pre:  0 <= x <= 0xFFFF.                                                    */
/* post: x + AddInv(x) == 0, where '+' is addition in the additive group.     */

#define AddInv(x)  (-x & ones)

/******************************************************************************/
/* Inverts a decryption/encrytion key to a encrytion/decryption key.          */
/* pre:  'key'    contains the encryption/decryption key.                     */
/* post: 'invKey' contains the decryption/encryption key.                     */

void _mcrypt_Idea_InvertKey(IDEA_KEY * key)
{
	register word16 t;
	register int lo, hi, i;

	lo = 0;
	hi = 6 * Idea_nofRound;
	t = MulInv(key->Idea_Key[lo]);
	key->Idea_inverted_Key[lo++] = MulInv(key->Idea_Key[hi]);
	key->Idea_inverted_Key[hi++] = t;
	t = AddInv(key->Idea_Key[lo]);
	key->Idea_inverted_Key[lo++] = AddInv(key->Idea_Key[hi]);
	key->Idea_inverted_Key[hi++] = t;
	t = AddInv(key->Idea_Key[lo]);
	key->Idea_inverted_Key[lo++] = AddInv(key->Idea_Key[hi]);
	key->Idea_inverted_Key[hi++] = t;
	t = MulInv(key->Idea_Key[lo]);
	key->Idea_inverted_Key[lo++] = MulInv(key->Idea_Key[hi]);
	key->Idea_inverted_Key[hi] = t;
	for (i = (Idea_nofRound - 1) / 2; i != 0; i--) {
		t = key->Idea_Key[lo];
		key->Idea_inverted_Key[lo++] = key->Idea_Key[hi -= 5];
		key->Idea_inverted_Key[hi++] = t;
		t = key->Idea_Key[lo];
		key->Idea_inverted_Key[lo++] = key->Idea_Key[hi];
		key->Idea_inverted_Key[hi] = t;
		t = MulInv(key->Idea_Key[lo]);
		key->Idea_inverted_Key[lo++] =
		    MulInv(key->Idea_Key[hi -= 5]);
		key->Idea_inverted_Key[hi++] = t;
		t = AddInv(key->Idea_Key[lo]);
		key->Idea_inverted_Key[lo++] = AddInv(key->Idea_Key[++hi]);
		key->Idea_inverted_Key[hi--] = t;
		t = AddInv(key->Idea_Key[lo]);
		key->Idea_inverted_Key[lo++] = AddInv(key->Idea_Key[hi]);
		key->Idea_inverted_Key[hi++] = t;
		t = MulInv(key->Idea_Key[lo]);
		key->Idea_inverted_Key[lo++] = MulInv(key->Idea_Key[++hi]);
		key->Idea_inverted_Key[hi] = t;
	}
#if (Idea_nofRound % 2 == 0)
	t = key->Idea_Key[lo];
	key->Idea_inverted_Key[lo++] = key->Idea_Key[hi -= 5];
	key->Idea_inverted_Key[hi++] = t;
	t = key->Idea_Key[lo];
	key->Idea_inverted_Key[lo++] = key->Idea_Key[hi];
	key->Idea_inverted_Key[hi] = t;
	key->Idea_inverted_Key[lo] = MulInv(key->Idea_Key[lo]);
	lo++;
	t = AddInv(key->Idea_Key[lo]);
	key->Idea_inverted_Key[lo] = AddInv(key->Idea_Key[lo + 1]);
	lo++;
	key->Idea_inverted_Key[lo++] = t;
	key->Idea_inverted_Key[lo] = MulInv(key->Idea_Key[lo]);
#else
	key->Idea_inverted_Key[lo] = key->Idea_Key[lo];
	lo++;
	key->Idea_inverted_Key[lo] = key->Idea_Key[lo];
#endif
}				/* Idea_InvertKey */

/******************************************************************************/
/* Expands a user key of 128 bits to a full encryption key                    */
/* pre:  'userKey' contains the 128 bit user key                              */
/* post: 'key'     contains the encryption key                                */

int _mcrypt_set_key(IDEA_KEY * key, Idea_UserKey userKey, int len)
{
	register int i;

#if (Idea_keyLen <= Idea_userKeyLen)
# ifdef WORDS_BIGENDIAN
	for (i = 0; i < Idea_keyLen; i++)
		key->Idea_Key[i] = byteswap16(userKey[i]);
# else
	for (i = 0; i < Idea_keyLen; i++)
		key->Idea_Key[i] = userKey[i];
# endif
#else
# ifdef WORDS_BIGENDIAN
	for (i = 0; i < Idea_userKeyLen; i++)
		key->Idea_Key[i] = byteswap16(userKey[i]);
# else
	for (i = 0; i < Idea_userKeyLen; i++)
		key->Idea_Key[i] = userKey[i];
# endif
	for (i = Idea_userKeyLen; i < Idea_keyLen; i++)
		if ((i & 7) < 6)
			key->Idea_Key[i] =
			    (key->Idea_Key[i - 7] & 127) << 9 | key->
			    Idea_Key[i - 6] >> 7;
		else if ((i & 7) == 6)
			key->Idea_Key[i] =
			    (key->Idea_Key[i - 7] & 127) << 9 | key->
			    Idea_Key[i - 14] >> 7;
		else
			key->Idea_Key[i] =
			    (key->Idea_Key[i - 15] & 127) << 9 | key->
			    Idea_Key[i - 14] >> 7;
#endif

	_mcrypt_Idea_InvertKey(key);

	return 0;
}				/* Idea_ExpandUserKey */

/******************************************************************************/


int _mcrypt_get_size()
{
	return sizeof(IDEA_KEY);
}

int _mcrypt_get_block_size()
{
	return 8;
}

int _is_block_algorithm()
{
	return 1;
}

int _mcrypt_get_key_size()
{
	return 16;
}

char *_mcrypt_get_algorithms_name()
{
	return "IDEA";
}

#define CIPHER "3223edc0f33ba078"

int _mcrypt_self_test()
{
	char *keyword;
	unsigned char *plaintext;
	unsigned char *ciphertext;
	int blocksize = _mcrypt_get_block_size(), j;
	void *key;
	unsigned char cipher_tmp[200];

	keyword = calloc(1, _mcrypt_get_key_size());
	for (j = 0; j < _mcrypt_get_key_size(); j++) {
		keyword[j] = ((j * 2 + 10) % 256);
	}
	ciphertext = calloc(1, blocksize);
	plaintext = calloc(1, blocksize);
	for (j = 0; j < blocksize; j++) {
		plaintext[j] = j % 256;
	}
	key = malloc(_mcrypt_get_size());
	memcpy(ciphertext, plaintext, blocksize);

	_mcrypt_set_key(key, (void *) keyword, _mcrypt_get_key_size());
	_mcrypt_encrypt(key, (void *) ciphertext);

	for (j = 0; j < blocksize; j++) {
		sprintf(&((char *) cipher_tmp)[2 * j], "%.2x",
			ciphertext[j]);
	}

	if (strcmp((char *) cipher_tmp, CIPHER) != 0) {
		printf("failed compatibility\n");
		printf("Expected: %s\nGot: %s\n", CIPHER,
		       (char *) cipher_tmp);
		return -1;
	}
	_mcrypt_decrypt(key, (void *) ciphertext);

	for (j = 0; j < blocksize; j++) {
		sprintf(&((char *) cipher_tmp)[2 * j], "%.2x",
			ciphertext[j]);
	}
	if (memcmp(ciphertext, plaintext, _mcrypt_get_block_size()) != 0) {
		printf("Got: %s\n", (char *) cipher_tmp);
		printf("failed internally\n");
		return -1;
	}
	return 0;
}

word32 _mcrypt_algorithm_version()
{
	return 20010701;
}
static const int key_sizes[] = { 16 };
const int *_mcrypt_get_supported_key_sizes(int *len)
{
	*len = sizeof(key_sizes)/sizeof(int);
	return key_sizes;
}
